pandas/_libs/internals.pyx

```diff
@@ -105,8 +105,7 @@ cdef class BlockPlacement:
             Py_ssize_t start, stop, end, _
         if not self._has_array:
             start, stop, step, _ = slice_get_indices_ex(self._as_slice)
-            self._as_array = np.arange(start, stop, step,
-                                       dtype=np.int64)
+            self._as_array = np.arange(start, stop, step, dtype=np.int64)
```

Comment:
some recent profiling has made me think this might be worth trying to optimize even harder.  two ideas come to mind:

1) use the numpy C-API to do this call faster
2) use range instead of np.arange (would end up being appreciably more invasive)

Reply:
@jbrockmendel absolutely (but non-cleanup PR)!